query-planner-js/src/QueryPlanningContext.ts

```diff
@@ -0,0 +1,304 @@
+        type && isEntityTypeMetadata(type)
+          ? type.keys.get(serviceName)
+          : undefined;
+
+      if (!(keys && keys.length > 0)) continue;
+
+      if (fetchAll) {
+        keyFields.push(
+          ...keys.flatMap(key =>
+            this.collectFields(scope.refine(possibleType), {
+              kind: Kind.SELECTION_SET,
+              selections: key,
+            }),
+          ),
+        );
+      } else {
+        keyFields.push(
+          ...this.collectFields(scope.refine(possibleType), {
+            kind: Kind.SELECTION_SET,
+            selections: keys[0],
+          }),
+        );
+      }
+    }
+
+    return keyFields;
+  }
+
+  getRequiredFields(
+    scope: Scope,
+    fieldDef: GraphQLField<any, any>,
+    serviceName: string,
+  ): FieldSet {
+    const requiredFields: FieldSet = [];
+
+    requiredFields.push(...this.getKeyFields(scope, serviceName));
+
+    const fieldFederationMetadata = getFederationMetadataForField(fieldDef);
+    if (fieldFederationMetadata?.requires) {
+      requiredFields.push(
+        ...this.collectFields(scope, {
+          kind: Kind.SELECTION_SET,
+          selections: fieldFederationMetadata.requires,
+        }),
+      );
+    }
+
+    return requiredFields;
+  }
+
+  getProvidedFields(
+    fieldDef: GraphQLField<any, any>,
+    serviceName: string,
+  ): FieldSet {
+    const returnType = getNamedType(fieldDef.type);
+    if (!isCompositeType(returnType)) return [];
+
+    const providedFields: FieldSet = [];
+
+    providedFields.push(...this.getKeyFields(Scope.create(this, returnType), serviceName, true));
+
+    const fieldFederationMetadata = getFederationMetadataForField(fieldDef);
+    if (fieldFederationMetadata?.provides) {
+      providedFields.push(
+        ...this.collectFields(Scope.create(this, returnType), {
+          kind: Kind.SELECTION_SET,
+          selections: fieldFederationMetadata.provides,
+        }),
+      );
+    }
+
+    return providedFields;
+  }
+
+  private getFragmentCondition(
+    parentType: GraphQLCompositeType,
+    fragment: FragmentDefinitionNode | InlineFragmentNode,
+  ): GraphQLCompositeType {
+    const typeConditionNode = fragment.typeCondition;
+    if (!typeConditionNode) return parentType;
+
+    return typeFromAST(
+      this.schema,
+      typeConditionNode,
+    ) as GraphQLCompositeType;
+  }
+
+  /**
+   * Given a current scope and a new type condition (fragment spread or inline fragment), computes the scope
+   * inside that new condition.
+   *
+   * Note that it is possible the new condition makes the branch unreachable (meaning that no possible
+   * runtime types can implement/be part of the previous type condition (in the scope) and the new one).
+   * In that case this method returns undefined and the branch should be ignored.
+   */
+  private scopeForFragment(
+    currentScope: Scope,
+    fragment: FragmentDefinitionNode | InlineFragmentNode,
+    fragmentDirectives?: ReadonlyArray<DirectiveNode>
```

Comment:
I had to check to see why we didn't just take `directives` from `fragment`, but I now understand these are the directives that are applied either on the `InlineFragmentNode` or the `FragmentSpreadNode` (so not the `FragmentNode`). Maybe `appliedDirectives` would be more clear? And/or a comment?